tsfc/interpreter.py

```diff
@@ -0,0 +1,306 @@
+    assert b.arr.dtype == numpy.dtype("bool")
+    result = Result.empty(a, b, dtype=bool)
+    for idx in numpy.ndindex(result.tshape):
+        result[idx] = a[a.filter(idx, result.fids)] or \
+            b[b.filter(idx, result.fids)]
+    return result
+
+
+@_evaluate.register(gem.Conditional)  # noqa: not actually redefinition
+def _(e, self):
+    cond, then, else_ = [self(o) for o in e.children]
+    assert cond.arr.dtype == numpy.dtype("bool")
+    result = Result.empty(cond, then, else_)
+    for idx in numpy.ndindex(result.tshape):
+        if cond[cond.filter(idx, result.fids)]:
+            result[idx] = then[then.filter(idx, result.fids)]
+        else:
+            result[idx] = else_[else_.filter(idx, result.fids)]
+    return result
+
+
+@_evaluate.register(gem.Indexed)  # noqa: not actually redefinition
+def _(e, self):
+    """Indexing maps shape to free indices"""
+    val = self(e.children[0])
+    fids = tuple(i for i in e.multiindex if isinstance(i, gem.Index))
+
+    idx = []
+    # First pick up all the existing free indices
+    for _ in val.fids:
+        idx.append(Ellipsis)
+    # Now grab the shape axes
+    for i in e.multiindex:
+        if isinstance(i, gem.Index):
+            # Free index, want entire extent
+            idx.append(Ellipsis)
+        elif isinstance(i, gem.VariableIndex):
+            try:
+                # Variable indices must be provided in bindings
+                idx.append(self.bindings[i])
+            except KeyError:
+                raise ValueError("Binding for %s not found" % i)
+        else:
+            # Fixed index, just pick that value
+            idx.append(i)
+    assert len(idx) == len(val.tshape)
+    return Result(val[idx], val.fids + fids)
+
+
+@_evaluate.register(gem.ComponentTensor)  # noqa: not actually redefinition
+def _(e, self):
+    """Component tensors map free indices to shape."""
+    val = self(e.children[0])
+    axes = []
+    fids = []
+    # First grab the free indices that aren't bound
+    for a, f in enumerate(val.fids):
+        if f not in e.multiindex:
+            axes.append(a)
+            fids.append(f)
+    # Now the bound free indices
+    for i in e.multiindex:
+        axes.append(val.fids.index(i))
+    # Now the existing shape
+    axes.extend(range(len(val.fshape), len(val.tshape)))
+    return Result(numpy.transpose(val.arr, axes=axes),
+                  tuple(fids))
+
+
+@_evaluate.register(gem.IndexSum)  # noqa: not actually redefinition
+def _(e, self):
+    """Index sums reduce over the given axis."""
+    val = self(e.children[0])
+    idx = val.fids.index(e.index)
+    return Result(val.arr.sum(axis=idx),
+                  val.fids[:idx] + val.fids[idx+1:])
+
+
+@_evaluate.register(gem.ListTensor)  # noqa: not actually redefinition
+def _(e, self):
+    """List tensors just turn into arrays."""
+    ops = [self(o) for o in e.children]
+    assert all(ops[0].fids == o.fids for o in ops)
```

Comment:
This is true for ListTensors where were translated from UFL. GEM doesn't have this restriction though.